File: app/src/main/java/app/we/go/emojidraw/widget/DrawingViewWithControls.java
package app.we.go.emojidraw.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.AccelerateInterpolator;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.List;

import app.we.go.emojidraw.R;
import app.we.go.emojidraw.model.Stroke;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

/**
 * Wraps the {@link DrawingView} and adds Undo, Clear and Skip buttons.
 */
@SuppressWarnings("WeakerAccess")
public class DrawingViewWithControls extends RelativeLayout {

    @BindView(R.id.drawing_area)
    DrawingView drawingView;

    @BindView(R.id.win_emoji)
    TextView winEmojiTextView;

    @BindView(R.id.win_overlay)
    View winOverlayView;


    private final PublishSubject<Object> skipPubSub = PublishSubject.create();

    // An irrelevant object that is emitted by the skipObservable
    @SuppressWarnings("FieldCanBeLocal")
    private Object irrelevant = new Object();


    public DrawingViewWithControls(final Context context) {
        super(context);
        init(context);
    }

    public DrawingViewWithControls(final Context context, final AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public DrawingViewWithControls(final Context context, final AttributeSet attrs, final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(final Context context) {
        inflate(context, R.layout.drawing_view_with_controls, this);
        ButterKnife.bind(this);
    }

    /**
     * @see DrawingView#getStrokesObservable()
     * @return
     */
    public Observable<List<Stroke>> getStrokesObservable() {
        return drawingView.getStrokesObservable();
    }

    /**
     * An observable that emits when the skip button is clicked
     * @return
     */
    public Observable<Object> getSkipObservable() {
        return skipPubSub;
    }


    @OnClick(R.id.undo_button)
    public void onUndo() {
        drawingView.undo();
    }

    @OnClick(R.id.clear_button)
    public void onClear() {
        drawingView.clear();
    }

    @OnClick(R.id.skip_button)
    public void onSkip() {
        drawingView.clear();
        irrelevant = new Object();
        skipPubSub.onNext(irrelevant);
    }


    public void animateWin(String winEmoji) {
        float scale = 1.75f;
        int duration = 500;
        winEmojiTextView.setAlpha(0.5f);
        winEmojiTextView.setText(winEmoji);
        winOverlayView.setVisibility(VISIBLE);
        winEmojiTextView.animate()
                .alpha(1f)
                .scaleX(scale)
                .scaleY(scale)
                .setInterpolator(new AccelerateInterpolator())
                .setDuration(duration)
                .withEndAction(() -> {
                    winOverlayView.setVisibility(GONE);
                    winEmojiTextView.setAlpha(0f);
                })
                .start();
    }
}

File: app/src/main/java/app/we/go/emojidraw/widget/DrawingViewWithControls.kt
package app.we.go.emojidraw.widget

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.animation.AccelerateInterpolator
import android.widget.RelativeLayout
import app.we.go.emojidraw.R
import app.we.go.emojidraw.model.Stroke
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject
import kotlinx.android.synthetic.main.drawing_view_with_controls.view.*

/**
 * Wraps the [DrawingView] and adds Undo, Clear and Skip buttons.
 */
class DrawingViewWithControls @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0, defStyleRes: Int = 0
) : RelativeLayout(context, attrs, defStyleAttr, defStyleRes) {

    private val skipPubSub = PublishSubject.create<Any>()

    init {
        View.inflate(context, R.layout.drawing_view_with_controls, this)

        undoButton.setOnClickListener {
            drawingArea.undo()
        }

        clearButton.setOnClickListener { drawingArea.clear() }

        skipButton.setOnClickListener {
            drawingArea.clear()
            skipPubSub.onNext(Any())
        }
    }


    /**
     * @see DrawingView.getStrokesObservable
     * @return
     */
    val strokesObservable: Observable<List<Stroke>>
        get() = drawingArea.strokesObservable

    /**
     * An observable that emits when the skip button is clicked
     * @return
     */
    val skipObservable: Observable<Any>
        get() = skipPubSub


    fun clear() {
        drawingArea.clear()
    }


    fun animateWin(winEmoji: String) {
        val scale = 1.75f
        val duration = 500
        winOverlay.visibility = View.VISIBLE
        winEmojiTextView.apply {
            alpha = 0.5f
            text = winEmoji
            animate()
                .alpha(1f)
                .scaleX(scale)
                .scaleY(scale)
                .setInterpolator(AccelerateInterpolator())
                .setDuration(duration.toLong())
                .withEndAction {
                    this@DrawingViewWithControls.winOverlay.visibility = View.GONE
                    winEmojiTextView.alpha = 0f
                }
                .start()
        }
    }
}
